### PR TITLE
Support testing without an application template wrapper.

### PR DESCRIPTION
**Pull request by cibernox**
Is this the best way to check if a feature flag is enabled?

### CODE REVIEW

File: addon-test-support/@ember/test-helpers/setup-application-context.js [hunk 1]
```diff
@@ -19,7 +19,12 @@ export function visit() {
       return owner.visit(...arguments);
     })
     .then(() => {
-      context.element = document.querySelector('#ember-testing > .ember-view');
+      // eslint-disable-next-line
+      if (EmberENV.FEATURES['ember-glimmer-remove-application-template-wrapper']) {
```
Comment (rwjblue) [inline addon-test-support/@ember/test-helpers/setup-application-context.js]: This might work for now, but isn't a good long term solution (once it moves to beta this will be falsey).

I think what we need to do is render an empty template, detect if there are any `ember-view`s and then memoize (so we only have to do the extra rendering one time for the entire suite)/

Comment (rwjblue) [reply addon-test-support/@ember/test-helpers/setup-application-context.js]: @cibernox - Now that this has settled, we can use `EmberENV._APPLICATION_TEMPLATE_WRAPPER`. A `true` value means we need `#ember-testing > .ember-view` and a `false` value means we need `#ember-testing`.

File: addon-test-support/@ember/test-helpers/setup-rendering-context.js [hunk 2]
```diff
@@ -196,7 +196,12 @@ export default function setupRenderingContext(context) {
       // In older Ember versions (2.4) the element itself is not stable,
       // and therefore we cannot update the `this.element` until after the
       // rendering is completed
-      context.element = getRootElement().querySelector('.ember-view');
+      // eslint-disable-next-line
+      if (EmberENV.FEATURES['ember-glimmer-remove-application-template-wrapper']) {
```
Comment (rwjblue) [inline addon-test-support/@ember/test-helpers/setup-rendering-context.js]: See above, should be able to use `EmberENV._APPLICATION_TEMPLATE_WRAPPER` here.

File: tests/dummy/config/environment.js [hunk 3]
```diff
@@ -11,6 +11,7 @@ module.exports = function(environment) {
       FEATURES: {
         // Here you can enable experimental features on an ember canary build
         // e.g. 'with-controller': true
+        'ember-glimmer-remove-application-template-wrapper': true,
```
Comment (rwjblue) [inline tests/dummy/config/environment.js]: Lets remove this.

In order to test (until https://github.com/ember-cli/ember-try/pull/181 lands in a release) we will need to:

1. add an ember-try scenario for `ember-canary-without-application-wrapper`
2. add that scenario to .travis.yml
3. update `tests/dummy/config/environment.js` to do:

```js
EmberENV: {
  // ...snip other normal stuff...
  _APPLICATION_TEMPLATE_WRAPPER: process.env.EMBER_TRY_CURRENT_SCENARIO === 'ember-canary-without-application-wrapper' ? false : true
}
```

File: addon-test-support/@ember/test-helpers/setup-application-context.js (hunk 1 above)
Comment (cibernox) [reply addon-test-support/@ember/test-helpers/setup-application-context.js]: I'll update it.

**Comment by cibernox** [comment]
@rwjblue I think this is ready.
However, how some the feature flag has become an underscored prop in `EmberENV`. Is this FF still enabled the same way?

**Comment by rwjblue** [comment]
> However, how some the feature flag has become an underscored prop in EmberENV. Is this FF still enabled the same way?

It is expected to be enabled via the `@ember/optional-features` addon which takes care of setting up the private environment flag. In the long run, all of this will be moved to be build time (so we don't ship both sets of implementation in Ember) but for now it is runtime (based on `EmberENV._APPLICATION_TEMPLATE_WRAPPER`).  

I'll update this addon in a followup PR to leverage `@ember/optional-features` and `ember-try`'s integration...

**Comment by cibernox** [comment]
@rwjblue Do the failure make sense to you?

**Comment by rwjblue** [comment]
Yes, I am working on fixing the canary failure in https://github.com/emberjs/ember.js/pull/16232. 

Also, I pushed some commits which cleanup the testing harness a bit (to more of what we expect "normal" addons to do going forward).

**Comment by rwjblue** [comment]
ZOMG, finally green.

**Comment by cibernox** [comment]
Looks great. :shipit:

File: .travis.yml [hunk 4]
```diff
@@ -44,6 +44,7 @@ jobs:
     - env: EMBER_TRY_SCENARIO=ember-beta
     - env: EMBER_TRY_SCENARIO=ember-canary
     - env: EMBER_TRY_SCENARIO=ember-default-with-jquery
+    - env: EMBER_TRY_SCENARIO=ember-canary-without-application-wrapper
```
Comment (Turbo87) [inline .travis.yml]: can we drop the `canary` from the scenario name? I assume this functionality will eventually exist outside of canary and that currently it uses canary is only an implementation detail, right?

File: addon-test-support/@ember/test-helpers/setup-application-context.js [hunk 5]
```diff
@@ -19,7 +19,12 @@ export function visit() {
       return owner.visit(...arguments);
     })
     .then(() => {
-      context.element = document.querySelector('#ember-testing > .ember-view');
+      // eslint-disable-next-line
```
Comment (Turbo87) [inline addon-test-support/@ember/test-helpers/setup-application-context.js]: I would prefer `/* global EmberENV */` because `eslint-disable-next-line` disables **all** linting rules

File: addon-test-support/@ember/test-helpers/setup-rendering-context.js [hunk 6]
```diff
@@ -196,7 +196,12 @@ export default function setupRenderingContext(context) {
       // In older Ember versions (2.4) the element itself is not stable,
       // and therefore we cannot update the `this.element` until after the
       // rendering is completed
-      context.element = getRootElement().querySelector('.ember-view');
+      // eslint-disable-next-line
```
Comment (Turbo87) [inline addon-test-support/@ember/test-helpers/setup-rendering-context.js]: see above

File: package.json [hunk 7]
```diff
@@ -54,9 +55,9 @@
     "ember-maybe-import-regenerator-for-testing": "^1.0.0",
     "ember-native-dom-event-dispatcher": "^0.6.3",
     "ember-resolver": "^4.0.0",
-    "ember-source": "~2.18.0",
+    "ember-source": "~3.0.0-beta.5",
     "ember-source-channel-url": "^1.0.1",
-    "ember-try": "^0.2.23",
+    "ember-try": "^1.0.0-beta.1",
```
Comment (Turbo87) [inline package.json]: offtopic: mmhhh.... I have thoughts about promoting this to 1.0.0...

File: tests/integration/setup-rendering-context-test.js [hunk 8]
```diff
@@ -119,18 +119,10 @@ module('setupRenderingContext "real world"', function(hooks) {
...
-        this.element,
-        'precond - confirm that the rootElement is different from this.element'
-      );
-
       this.set('rootElement', rootElement);
 
       await render(hbs`{{#-in-element rootElement}}{{click-me-button}}{{/-in-element}}`);
 
-      // this will need to be modified / removed once RFC280 lands
-      assert.equal(this.element.textContent, '', 'no content is contained _within_ this.element');
```
Comment (Turbo87) [inline tests/integration/setup-rendering-context-test.js]: why is this removed?

File: tests/integration/setup-rendering-context-test.js [hunk 9]
```diff
@@ -119,18 +119,10 @@ module('setupRenderingContext "real world"', function(hooks) {
     test('can click on a sibling element of this.element (within the rootElement)', async function(assert) {
       let rootElement = document.getElementById('ember-testing');
 
-      assert.notEqual(
-        rootElement,
-        this.element,
-        'precond - confirm that the rootElement is different from this.element'
-      );
```
Comment (Turbo87) [inline tests/integration/setup-rendering-context-test.js]: should we keep this assertion and run it conditionally instead?

File: package.json (hunk 7 above)
Comment (rwjblue) [reply package.json]: haha, we can always make a 2.0 😸 If there are specific things wrong with ember-try, please make issues so we can discuss the details.

IMHO, there is basically no reason to not have the ability to signal "new API's are added" (normal minor version bump) from "bug fixes" (normal patch release).

File: tests/integration/setup-rendering-context-test.js (hunk 8 above)
Comment (rwjblue) [reply tests/integration/setup-rendering-context-test.js]: Because `this.element` (when the optional flag is enabled) _will_ have content. I can swap things around a bit and guard the assertions, but I think I'd rather have two tests instead.  One for "with a wrapper div" and one "without a wrapper"...

File: tests/integration/setup-rendering-context-test.js (hunk 9 above)
Comment (rwjblue) [reply tests/integration/setup-rendering-context-test.js]: ya, I'll make it two tests

File: package.json (hunk 7 above)
Comment (Turbo87) [reply package.json]: https://github.com/ember-cli/ember-try/issues/155

File: tests/integration/setup-rendering-context-test.js (hunk 8 above)
Comment (Turbo87) [reply tests/integration/setup-rendering-context-test.js]: what content will be in there and why wasn't it in there before? I'm not sure I understand the assertion 🤔

Comment (rwjblue) [reply tests/integration/setup-rendering-context-test.js]: The test (as it exists on master) is actually testing a few different things:

* `this.element` within the test context is **not** the `#ember-testing` div
* when rendering _only_ content that is "wormholed" (via `in-element` in this case), the actual `this.element` is still empty (because the wormholed content is somewhere else)
* that we _can_ interact (e.g. click) with elements that are within the `#ember-testing` div (even though they are above the `this.element`).

As you can tell some of these assertions are no longer needed with the changes from emberjs/rfcs#280.  Specifically:

> `this.element` within the test context is **not** the `#ember-testing` div

without the wrapper DIV, the `this.element` actually _is_ `#ember-testing`

>  when rendering _only_ content that is "wormholed", the actual `this.element` is still empty

due to the above, this is not true any longer. `this.element` will have content (the wormholed content)

These two changes are the assertions that were being removed (when you commented). I have changed things around a bit so that we have two different tests (one with an application template wrapper and one without it).

**Comment by rwjblue** [comment]
Created https://github.com/emberjs/ember-optional-features/issues/10 to track swapping out the manual `EmberENV` checks in favor of something more reliable provided by `@ember/optional-features`.

File: tests/integration/setup-rendering-context-test.js (hunk 8 above)
Comment (Turbo87) [reply tests/integration/setup-rendering-context-test.js]: > that we can interact (e.g. click) with elements that are within the #ember-testing div (even though they are above the this.element).

given that we use `document.querySelector('#ember-testing > .ember-view')` how is `click()` able to interact with elements above `this.element`? last time I've tried it didn't work with wormholed content (yet) 🤔

Comment (rwjblue) [reply tests/integration/setup-rendering-context-test.js]: This behavior was changed in https://github.com/emberjs/ember-test-helpers/pull/295. `click` and friends are always operating on the root element (`#ember-testing` generally speaking) which is **not** `this.element`.  The specific test that we are commenting on ATM is actually introduce _in_ that PR specifically to address the wormholed content scenario...

Comment (Turbo87) [reply tests/integration/setup-rendering-context-test.js]: I see. Turns out we are still on v0.7.10 of `@ember/test-helpers` (though newest `ember-cli-qunit`) which is the reason why it didn't work when I tried it.
